crowbar_framework/app/controllers/api/restart_management_controller.rb

```diff
@@ -0,0 +1,144 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+class Api::RestartManagementController < ApiController
+  skip_before_filter :upgrade
+
+  api :POST, "/api/restart_management/configuration", "Set the disallow restart value for barclamps"
+  header "Accept", "application/vnd.crowbar.v2.0+json", required: true
+  param :disallow_restart, [true, false], desc: "Disallow service reboots", required: true
+  param :barclamp, String, desc: "Barclamp to apply to", required: true
+  api_version "2.0"
+
+  api :GET, "/api/restart_management/configuration", "List the disallow restart value for barclamps"
+  header "Accept", "application/vnd.crowbar.v2.0+json", required: true
+  api_version "2.0"
+  def configuration
+    if request.post?
+      configuration_post
+    else
+      configuration_get
+    end
+  end
+
+  api :POST, "/api/restart_management/restarts", "Clean the service restart flags"
+  header "Accept", "application/vnd.crowbar.v2.0+json", required: true
+  param :node, String, desc: "Node name", required: true
+  param :service, String, desc: "Service to clean restart flag for", required: true
+  api_version "2.0"
+
+  api :GET, "/api/restart_management/restarts", "Get a list of services that need restart"
+  header "Accept", "application/vnd.crowbar.v2.0+json", required: true
+  api_version "2.0"
+  def restarts
+    if request.post?
+      restarts_post
+    else
+      restarts_get
+    end
+  end
+
+  private
+
+  # Gets a list of the barclamps and extract the values from their nodes
+  # this ignores the roles and groups together all the nodes under a barclamp
+  # this is because we use a barclamp-level service restart checks instead of
+  # doing it service level, which is more cumbersome and requires us to know a list of services
+  # beforehand
+  def restarts_get
+    # Chef query search to find all nodes that have requires_restart attribute not empty
+    nodes_with_restarts = NodeObject.find("requires_restart:*")
+    restart_requests_per_node = {}
+    nodes_with_restarts.each do |node|
+      restart_requests_per_node[node[:fqdn]] = { alias: node.alias }
+      managed_barclamps.each do |barclamp|
+        # skip if there is not attributes for the barclamp or the requires_restart is empty
+        # why node.node.fetch? Dont ask me, ask the chef creators why this works
+        # https://github.com/SUSE-Cloud/chef/blob/10-stable-suse/chef/lib/chef/node.rb#L187
+        requires_restart = node.node.fetch(
```

Comment:
`node.node`? Note that you can use `node.crowbar_wall.fetch(...` as `crowbar_wall` is a method that does what you want.

Reply:
+1

Reply:
Interesting that changing this makes the test fail as crowbar_wall is not returning the expected values. Not sure if its due the test environment and out fake nodes, or it just has a different approach

Reply:
Disregard the above comment. I did not change the fetching correctly so of course it didnt work. Duh